doc/design/clustered-etcd-alert-store/design.md

```diff
@@ -0,0 +1,150 @@
+# Clustered Etcd Alert Store
+
+Type: Design Document (and PR)
+
+Date: 2019-09-20
+
+Author: David C Wang <dcwangmit01@gmail.com>
+
+## Status Quo
+
+The current Alertmanager clustering implementation will allow multiple
+instances of Alertmanager to sync recently-updated notification log entries and
+alert silences via the gossip protocol.  The syncing of alert state is not yet
+implemented.
+
+Currently, Alertmanager alert state is entirely stored in process memory.  It
+is lost upon restart of Alertmanager.  This, in conjunction with the fact that
+Alertmanager alert state is not synced with clustered peers presents several
+problems:
+
+* Alert emitters must send alerts to every single Alertmanager instance in the
+  cluster.  If some Alertmanager instances receive alerts that other instances
+  do not, their state will drift apart.
+
+  * Typically, the alert emitter is Prometheus which uses service discovery to
+    locate all Alertmanager instances, and then sends alerts to each individual
+    target.  Custom alert emitters may not have that service discovery
+    capability and must be explicity configured.
+
+* Because the alert must be sent to every Alertmanager instance, this precludes
+  a typical HA deployment pattern of having clients (i.e. alert emitters) send
+  alerts through a Load Balancer that fronts several Alertmanager instances.
+  This pattern is typical in Kubernetes.
+
+* Restarting Alertmanager instances or increasing the replicaCount for the
+  Alertmanager cluster will result in the new members having very different
+  alert states than current members.
```

Comment:
> Restarting Alertmanager instances

When each instance has persistence they should have a *good-enough* snapshot of the current notification and silcence state on disk. Thereby end-users should not have duplicate notifications.

Reply:
It is not the state of notifications or silences that is the concern of this PR.  This PR mostly involves the internal state of the alerts which are captured in the *mem* provider.  In our situation receiving duplicate notifications is okay.  What we have trouble with is Alertmanager losing its internal alert state and never ever sending notifications to close alerts that are no longer active.